src/artifacts.js

```diff
@@ -423,6 +643,119 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
   this.monitor.reportError(err);
 };
 
+/** Complete artifact */
+api.declare({
+  method:     'put',
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'completeArtifact',
+  stability:  API.stability.experimental,
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>',
+    ], [
+      'queue:create-artifact:<taskId>/<runId>',
+    ],
+  ],
+  deferAuth:  true,
+  input:      'put-artifact-request.json#',
+  output:     'put-artifact-response.json#',
+  title:      'Complete Artifact',
+  description: 'tbd',
+}, async function(req, res) {
+  let taskId      = req.params.taskId;
+  let runId       = parseInt(req.params.runId, 10);
+  let name        = req.params.name;
+  let input       = req.body;
+  let isPublic    = /^public\//.test(name);
+
+  // Hmm, list destructuring is probably better here
+  let [artifact, task] = await Promise.all([
+    this.Artifact.load({taskId, runId, name}, true),
+    this.Task.load({taskId}, true),
+  ]);
+
+  // TODO: Have a shortcut here where things which are already
+  // marked as present short circuit
+
+  let run = task.runs[runId];
+  if (!run) {
+    return res.reportError('InputError',
+      'Run not found',
+      {});
+  }
+  let workerGroup = run.workerGroup;
+  let workerId = run.workerId;
+
+  // Authenticate request by providing parameters
+  if (!req.satisfies({
+    taskId,
+    runId,
+    workerGroup,
+    workerId,
+    name,
+  })) {
+    return;
+  }
+
+  if (artifact.storageType === 'blob') {
+    let etag;
+    if (artifact.details.partsHash) {
+      etag = await this.s3Controller.completeMultipartUpload({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        etags: input.etags,
+        uploadId: artifact.details.uploadId,
+        tags: {taskId, runId, name},
+      });
+    } else {
+      let url = await this.s3Controller.generateGetUrl({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        signed: !isPublic,
+      });
+
+      if (urllib.parse(url).protocol !== 'https:') {
+        throw new Error('Only HTTPS is allowed for artifacts');
+      }
+
+      let headRes = await this.s3Runner.run({
+        req:{
+          url,
+          method: 'HEAD',
+          headers: {},
+        },
+      });
+      
+      if (headRes.headers['x-amz-meta-content-sha256'] !== artifact.details.contentSha256) {
+        throw new Error('S3 object does not have correct Content-Sha256 metadata');
```

Comment:
Maybe add a comment explaining why this should not be possible... Perhaps even insert a label into the error indicating that this could be a sign of an attack... right? I mean, this would sort of only be possible if someone was actively trying to insert invalid artifacts...

Reply:
I don't know that it strictly implies an attack, rather it just shows that someone is uploading incorrectly.  What would be a stronger indication of an attack is that the header and artifact.details.contentSha256 values are the same, but the sha256 of the artifact as stored in S3 doesn't match either.

This is more of a sanity check and doesn't give us the strong check that signing the requests to only work with a specific payload body does.

Reply:
Is `x-amz-meta-content-sha256` covered by the signature?

If **no** then this is a 4xx error, because the user changed headers from the upload request.

if **yes** then a request upload this file should not be possible, the queue should never have signed it. Hence, someone changed the database, someone has S3 access, or we have a massive bug. Eitherway, it's serious as in we're under attack or have a massive flaw.

Reply:
@jhford, note: you can't throw an error if this isn't an internal error, if it can be provoked by the client throwing an error is a bug, as it results in internal error.